### 2.2 Question 2

Download the Iris dataset from https://archive.ics.uci.edu/ml/machine-learning-databases/iris/. The dataset can be downloaded from iris.data. Load the data into a pandas dataframe.
For this lab, we’re going to be performing a binary classification problem,
but this dataset has 3 classes: setosa, virginica, and versicolor. So we want
to take this multi-class problem and transform it into a binary classification.
Create a new column for the dataset called target. The value of target
will be 1 if the row contains a setosa flower, else the value is $0$. There should
be $\frac{1}{3}$ rows with the value of $1$, the rest should be $0$.

In [7]:
import pandas as pd
import numpy as np

# Defining the fielpath
filepath = "data\iris.data" 

def load_data(filepath):
    
    # This part reads the data and avoids to use the first row as the column name
    df = pd.read_csv(filepath, header = None)
    
    # After that the column names for the Dataset are assigned
    col_names = ['Sepal_Length','Sepal_Width','Petal_Length','Petal_Width','Species']
    df.columns = col_names  
    
    # Then the column "tareget" is created and filled with ceros
    df['target']  = 0
    
    # Then the values in "target" column are modified depending if the Specie is or not a Setosa
    df.loc[df['Species'] == 'Iris-setosa', 'target'] = 1
    
    return df

# The previous function is called
iris  = load_data(filepath)
iris

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width         Species  \
0             5.1          3.5           1.4          0.2     Iris-setosa   
1             4.9          3.0           1.4          0.2     Iris-setosa   
2             4.7          3.2           1.3          0.2     Iris-setosa   
3             4.6          3.1           1.5          0.2     Iris-setosa   
4             5.0          3.6           1.4          0.2     Iris-setosa   
..            ...          ...           ...          ...             ...   
145           6.7          3.0           5.2          2.3  Iris-virginica   
146           6.3          2.5           5.0          1.9  Iris-virginica   
147           6.5          3.0           5.2          2.0  Iris-virginica   
148           6.2          3.4           5.4          2.3  Iris-virginica   
149           5.9          3.0           5.1          1.8  Iris-virginica   

     target  
0         1  
1         1  
2         1  
3         1  
4         1  
..      ...  
145       0  
146       0  
147       0  
148       0  
149       0  

[150 rows x 6 columns]

### 2.3 Question 3
For this question we want to take this dataset of 150 rows, and split it into
a train, test, and validation dataset, using the following proportions for each
split:

• Training: 70%

• Validation: 10%

• Testing: 20%

Sample data for each subset using stratified sampling. I.e. the training data should have roughly $\frac{1}{3}$
positive samples, the testing and validation dataset should also have roughly $\frac{1}{3}$ positive samples.


In [8]:
iris['Species'].value_counts()

Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: Species, dtype: int64

In [9]:
def slipt_data(df, column):
    
    values = list(df[column].unique())
    
    training_set = pd.DataFrame()
    testing_set = pd.DataFrame()
    validation_set = pd.DataFrame()
    
    for value in values:
        
        data = df[df[column] == value]
        train=data.sample(n = 35,random_state=42)
        test=data.drop(train.index)
        validation = test.sample(n=5,random_state=42)
        test=test.drop(validation.index)
        
        training_set = pd.concat([training_set, train])
        testing_set = pd.concat([testing_set, test])
        validation_set = pd.concat([validation_set, validation])

    return training_set, testing_set, validation_set

In [10]:
training_set, testing_set, validation_set = slipt_data(iris, 'Species')

### 2.4 Question 4
Using the linear regression model you created in the previous lecture, transform it into a logistic regressor by applying the logistic function to the output
of the model. The loss function for this model should be binary cross entropy.

Select two columns from the Iris dataset (i.e. petal length and petal
width), and using these two columns, train a logistic regressor using gradient
descent, measuring the gradient using finite differences approximation. This
means that instead of having a single slope variable, we have multiple:

$$\hat{y} = \sigma\left (\beta_{0} + \sum_{i = 1}^{m}x_{i}\beta_{i}\right )$$


where $\hat{y}$ is the model’s probability prediction, $\sigma$ is the logistic/sigmoid
function, $\beta_{0}$ is the intercept, $\beta_{i}$
is the coefficient that modulates the $X_{i}$ variable.

I’ve made a start for you, please fill in the ’#TODOs’:

In [ ]:

def bce(y, yhat):
    # TODO: apply the binary cross entropy function returning the loss
    
    
    return loss
class LogisticRegressor:
    def __init__(self, n_features: int = 2):
    self.params = np.random.randn(n_features + 1)
    def logistic(self, x):
    # TODO: apply the logistic function
    return x
    def __call__(self, x, logits=False):
    y = self.params[0] + self.params[1:] @ x.T
    if not logits:
    y = self.logistic(y)
    return y
    def fit(train_x, train_y, valid_x, valid_y, epochs: int = 100, lr: float = 0.01):
    # TODO: train the model using gradient descent and finite-differences
    for epoch in range(1, epochs+1):
    for xi, yi in zip(train_x, train_y):
    # calculate loss and update model parameters using gradient descent
    for xi, yi in zip(valid_x, valid_y):
    # calculate validation loss (BUT DON'T UPDATE MODEL PARAMETERS!)
    def predict(x, logits):
    return self(x, logits=logits)

### 2.5 Question 5
As gradient descent is iterating, store (using class variables), the training and validation loss.

Visualise the training and validation loss. Is there a point at which the model begins to over fit? How do you know that the model is beginning to overfit by looking at these curves?

### 2.6 Question 6
Predict the class labels for the testing set.
For the testing set, calculate the:

• TP – number of true positives

• TN – number of true negatives

• FP – number of false positives

• FN – number of false negatives

In [5]:
def confusion_matrix(y, yhat):
    
    """
    This function returns the elements of a two-dimensional confusion matrix 
    calculated by counting the true values resulting of using logical operators such that:

        TP = Logical AND
        TN = Logical NOR
        FP = Logical A'B
        FN = Logical AB'

    """
    
    # Logical AND Gate
    TP = np.sum(np.logical_and(y, yhat))
    
    # Logical NOR Gate
    TN = np.sum(np.logical_and(np.logical_not(y),np.logical_not(yhat)))
    
    # Logical A'B Gate
    FP = np.sum(np.logical_and(np.logical_not(y),yhat))
    
    # Logical AB' Gate
    FN = np.sum(np.logical_and(y,np.logical_not(yhat)))
    
    return TP, TN, FP, FN

In [11]:
y =    np.array([1,1,0,1,0,0])
yhat = np.array([0,0,1,1,0,0])



TP, TN, FP, FN = confusion_matrix(y, yhat)

print("TP "+str(TP) + ",TN "+ str(TN) + ",FP "+ str(FP) + ",FN "+ str(FN) )

TP 1,TN 2,FP 1,FN 2


### ------------------------------------------

In [57]:
x = iris[["Sepal_Length", "Sepal_Width"]]

def linear_regresor(x):
    
    #np.random.seed(42)
    B0 = np.random.rand()
    coeficients = np.random.rand(x.shape[1])
    
    yhat = B0 + np.dot(x,coeficients)
    
    return yhat

In [58]:
def logistic(x):
    # TODO: apply the logistic function
    
    p = 1 / (1 + np.exp(-x))
    
    return p

In [59]:


def bce(y, yhat):
    # TODO: apply the binary cross entropy function returning the loss
    
    negative_side = (1-y)* np.log(1-logistic(yhat))
    positive_side = (y)* np.log(logistic(yhat))
    N = len(y)
    
    bce = (-1*np.sum(negative_side + positive_side )) / N
    
    return bce
    

In [62]:
(-1*np.sum(10 + 6 )) / 2

-8.0

In [65]:
-0.5 * np.sum(10 + 6 )

-8.0

In [60]:
bce(iris['target'],logistic(linear_regresor(x)))

0.9617694741336733

In [66]:
iris['target']

0      1
1      1
2      1
3      1
4      1
      ..
145    0
146    0
147    0
148    0
149    0
Name: target, Length: 150, dtype: int64

In [56]:
logistic(linear_regresor(x))

array([0.98302176, 0.97574377, 0.97721605, 0.97511628, 0.98354868,
       0.98776095, 0.97920236, 0.98144689, 0.97033653, 0.97715051,
       0.98618976, 0.98035602, 0.97504486, 0.97124561, 0.98973285,
       0.99170772, 0.98776095, 0.98302176, 0.98807347, 0.98582857,
       0.98345328, 0.98494787, 0.98155364, 0.98085397, 0.98035602,
       0.97642358, 0.98144689, 0.98350105, 0.98247827, 0.97721605,
       0.97649115, 0.98345328, 0.9885124 , 0.9900825 , 0.97715051,
       0.9790823 , 0.98486046, 0.97715051, 0.97204788, 0.98196982,
       0.9825288 , 0.95889839, 0.97518751, 0.9825288 , 0.98582857,
       0.97504486, 0.98582857, 0.97655854, 0.98578745, 0.98029933,
       0.98820807, 0.98598815, 0.98710797, 0.96896061, 0.98267491,
       0.97822845, 0.98642427, 0.96535742, 0.98414722, 0.97335998,
       0.95737673, 0.98176046, 0.9714018 , 0.98170775, 0.97890093,
       0.98634532, 0.9801283 , 0.97753503, 0.97297621, 0.97320711,
       0.98382731, 0.980576  , 0.97803985, 0.980576  , 0.98321

### 2.7 Question 7
Calculate the precision and recall and F1 score.


$$Precission = \frac{TP}{TP + FP}$$

$$Recall = \frac{TP}{TP + FN}$$

$$F_{\beta} = (1 + \beta^{2})\frac{Precission * Recall}{(\beta^{2}* Precision) + Recall)}$$

Calling the above function and making use of the previous equations we have that the metrics of the confusion matrix are given by:

In [ ]:

def precision(y, yhat):
    # calculate the precision and return it
    TP, TN, FP, FN = confusion_matrix(y, yhat)  
    pr = TP / (TP + FP)
    
    return pr

def recall(y, yhat):
    # calculate the recall and return it
    TP, TN, FP, FN = confusion_matrix(y, yhat)
    rc = TP / (TP + FN)
    
    return rc

def f_beta(y, yhat, beta=1):
    
    pr = precision(y, yhat)
    rc = recall(y, yhat)
    
    # calculate the f_beta score and return it
    fb = (1 + beta**2) * ((pr * rc) / ((beta**2 * pr) + rc))
    
    return fb


### 2.8 Question 8
Generate a report using the precision, recall and F1 and confusion matrix.
The report should be printed like:

In [ ]:
t1 = "|" + " " * 8 + "|" + " " * 10 + "| Predicted |" + " " *10 + "|\n"
t2 = "|" + " " * 8 + "|" + " " * 10 + "|  Positive |" + " Negative |\n"
t3 = "| Actual |" + " Positive |" + " " * (10 - len(str(TP))) + str(TP) + " |" + " " * (9 - len(str(FN))) + str(FN) + " |\n"
t4 = "|" + " " * 8 + "|" + " Negative |" + " " * (10 - len(str(FP))) + str(FP) + " |" + " " * (9 - len(str(TN))) + str(TN) + " |\n"

print(t1+t2+t3+t4)

### 2.9 Question 9
Calculate the true-positive and false positive rate, and from these values
generate a ROC curve.


In [ ]:
def roc(y, yhat, threshold_step=0.01):
# iteratively increase the threshold by threshold_step,
# calculating the TP and FP rate for each iteration. This function
# should return two lists, a list of TP rates, and a list of FP
# rates.
    return tp, fp

tp, fp = roc(y, yhat)

### 2.10 Question 10
Now that you’ve created a logistic classifier for two features of the Iris dataset
and have created some analytic results. Select another two columns (i.e.
petal width and sepal length, or petal length and sepal width). Create
a different logistic classifier using these new columns and create the same
results as you did with questions 8 and 9.
Compare these two models trained with different columns. Which model
is best, and why do we know that it’s the best?
